iroh-net/src/discovery.rs

```diff
@@ -148,6 +239,7 @@ where
     }
 }
 
+#[async_trait]
```

Comment:
I don't think we need this, we can just do `fn foo() -> impl Future` I think

Reply:
LMAO YES TY

Reply:
Okay, I've tried a few iterations of this.

When converting `subscribe` to use `fn subscribe(&self) -> impl Future`
I get an error saying that we cannot convert the discovery trait into an Object

And when I tried instead to use `BoxFuture` in the signature, I was getting errors attempting to return `None`. I tried returning `Box::pin(FutureOption::None)`, but that also gave me errors.

Can I get a hand here?

Appreciate any help.

Reply:
Why wasn't this a problem when using the async_trait macro? Because async_trait always just uses Box<dyn Future<Output = T>> under the hood. That is one of the main criticisms of the async_trait crate and why people were so happy about RPITIT.

About the next compile error: the fn now needs to return a Future<Option<Stream<T>>>, so the return type would have to be something like `Box::pin(future::ready(None))`.

But this signature is starting to become a bit annoying. I think the reason we returned Option<Stream> instead of Stream was that it would be easy and cheap to implement the trait with resolve disabled, by just returning None. No boxing, no nothing. Just None.

If you need to do something async when creating the stream, can't you already do this when constructing the stream? TBH I don't see the reason why this fn needs to be async.

Reply:
OK, to make this more clear: the most obvious signature for resolve would have been just this:

```rust
fn resolve(&self, node_id: NodeId) -> BoxStream<Result<DiscoveryItem>>
```

But then if you don't want to implement resolve, like a publish-only discovery impl, you would have to return `Box::pin(futures::stream::empty())` which seemed like an unnecessary box. That is why it returns an option. None just stands for the empty stream.

If you want to do some async setup before creating the stream, you can either use this pattern:

```rust
let res = async move {
  let foo = setup().await;
  make_stream(foo) // returns an impl Stream<Result<DiscoveryItem>>
}.flatten_stream()
Some(Box::pin(res))
```

or use genawaiter. So I am not sure what a signature of BoxFuture<Option<BoxStream>> buys you over just BoxStream or Option<BoxStream>.